src/coreclr/jit/optimizer.cpp

```diff
@@ -4564,6 +4564,299 @@ PhaseStatus Compiler::optUnrollLoops()
+                        return false;
+                    }
+
+                    asgNode = tree;
+                    asgStmt = stmt;
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    if (tree->gtGetOp1() != nullptr || (tree->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+    }
+    if (asgNode == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("Conditionally executing " FMT_BB " inside " FMT_BB "\n", middleBlock->bbNum, block->bbNum);
+        for (BasicBlock* dumpBlock = block; dumpBlock != middleBlock->bbNext; dumpBlock = dumpBlock->bbNext)
+        {
+            fgDumpBlock(dumpBlock);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    // Duplicate the input of the assign.
+    // This will be used as the false result of the select node.
+    GenTree* destination = gtCloneExpr(asgNode->AsOp()->gtOp1);
+    destination->gtFlags &= GTF_EMPTY;
+
+    // Invert the condition.
+    cond->gtOper  = GenTree::ReverseRelop(cond->gtOper);
+
+    // Create a select node.
+    GenTreeConditional* select =
+        gtNewConditionalNode(GT_SELECT, cond, asgNode->gtGetOp2(), destination, asgNode->TypeGet());
+    gtSetEvalOrder(select);
+
+    // Use the select as the source of the assignment.
+    asgNode->AsOp()->gtOp2 = select;
+    asgNode->AsOp()->gtFlags |= (select->gtFlags & GTF_ALL_EFFECT);
+    gtSetEvalOrder(asgNode);
+    fgSetStmtSeq(asgStmt);
+
+    // Remove the JTRUE statement.
+    last->ReplaceWith(gtNewNothingNode(), this);
+    gtSetEvalOrder(last);
+    fgSetStmtSeq(block->lastStmt());
+
+    // Move the Asg to the end of the original block
+    Statement* stmtList1 = block->firstStmt();
+    Statement* stmtList2 = middleBlock->firstStmt();
+    Statement* stmtLast1 = block->lastStmt();
+    Statement* stmtLast2 = middleBlock->lastStmt();
+    stmtLast1->SetNextStmt(stmtList2);
+    stmtList2->SetPrevStmt(stmtLast1);
+    stmtList1->SetPrevStmt(stmtLast2);
+    middleBlock->bbStmtList = nullptr;
+
+    // Fix up the SSA of assignment destination.
+    if (asgNode->AsOp()->gtOp1->IsLocal())
+    {
+        GenTreeLclVarCommon* lclVar = asgNode->AsOp()->gtOp1->AsLclVarCommon();
+        unsigned             lclNum = lclVar->GetLclNum();
+        unsigned             ssaNum = lclVar->GetSsaNum();
+
+        if (ssaNum != SsaConfig::RESERVED_SSA_NUM)
+        {
+            LclSsaVarDsc* ssaDef = lvaTable[lclNum].GetPerSsaData(ssaNum);
+            noway_assert(ssaDef->GetBlock() == middleBlock);
+            ssaDef->SetBlock(block);
```

Comment:
Doesn't this mean `LclSsaVarDsc::m_asg` is left pointing at an assignment that has been removed?

Updating that too will mean we end up with two SSA defs for the same assignment. I wonder if that will be a problem.

Reply:
I suppose for now we won't duplicate any SSA defs since this PR does not handle if-else. But, once that is handled, we will hit the situation. So for now things should be ok.

Also small nit: `lvaTable[lclNum] -> lvaGetDesc(lclNum)`

Reply:
> I think I just need to connect the uses somehow.

... which isn't easy as I'd essentially need to run an entire CSE pass across the block to be able to spot that node 88 and my new node 278 are the same.

I wonder if this should be done before the CSE phase. But then we need to add SELECT support to CSE and assert prop. Which is why the If conversion was placed after CSE.

Reply:
Note that even if this was before CSE, CSE currently doesn't handle `LCL_VAR` nodes (it wouldn't be hard to add, but may have some trailing regressions).

I'd echo Andy's suggestion to look into costing: https://github.com/dotnet/runtime/pull/73472/files#r942623060. E. g. one could have something like "estimated gain" (from removing the branch) vs "estimated loss" (from evaluating more expressions).

> `LCL_VAR   int    V00 loc0         u:15 $VN.Void`

Side note: it is not right for this use to have `$VN.Void`.

Reply:
>Note that even if this was before CSE, CSE currently doesn't handle LCL_VAR nodes (it wouldn't be hard to add, but may have some trailing regressions).

So what would ensure that this block doesn't load v00 twice?
```
*  ASG       int    $VN.Void
+--*  LCL_VAR   int    V00 loc0
\--*  ADD (or MUL, or whatever)
   +--*  LCL_VAR   int    V00 loc0
   \--*  LCL_VAR   int    V00 loc0 
```

>I'd echo Andy's suggestion to look into costing:

Right, because even without the above fixed, the costing will ensure we never regress in code size. I'll add something here.


Meanwhile, looking back to the queue testing. I was using Perf_PriorityQueue tests which uses a PriorityQueue and not a queue. 

I took those tests and rewrote them to use Queue: https://github.com/dotnet/performance/blob/1c5e29da0e745809bd911070bbef933d25328eaf/src/benchmarks/micro/libraries/System.Collections/Queue/Perf_Queue.cs

Results were much more stable, not really changing one run to the next:
```
| Dequeue_And_Enqueue | Job-BTPMBD | /runtime_HEAD/ |   10 |    124.2 ns |  0.17 ns |  0.16 ns |    124.2 ns |    123.9 ns |    124.4 ns |  1.00 |            Base |         - |          NA |
| Dequeue_And_Enqueue | Job-OHELEL | /runtime_csel/ |   10 |    130.7 ns |  0.79 ns |  0.66 ns |    130.8 ns |    129.7 ns |    132.0 ns |  1.05 |          Slower |         - |          NA |
| Dequeue_And_Enqueue | Job-BTPMBD | /runtime_HEAD/ |  100 |  1,219.6 ns |  0.55 ns |  0.46 ns |  1,219.8 ns |  1,218.7 ns |  1,220.1 ns |  1.00 |            Base |         - |          NA |
| Dequeue_And_Enqueue | Job-OHELEL | /runtime_csel/ |  100 |  1,200.3 ns |  4.17 ns |  3.69 ns |  1,199.4 ns |  1,196.5 ns |  1,206.3 ns |  0.98 |            Same |         - |          NA |
| Dequeue_And_Enqueue | Job-BTPMBD | /runtime_HEAD/ | 1000 | 11,892.0 ns |  4.18 ns |  3.49 ns | 11,891.9 ns | 11,887.0 ns | 11,899.7 ns |  1.00 |            Base |         - |          NA |
| Dequeue_And_Enqueue | Job-OHELEL | /runtime_csel/ | 1000 | 11,776.6 ns | 41.06 ns | 34.29 ns | 11,758.3 ns | 11,745.8 ns | 11,835.3 ns |  0.99 |            Same |         - |          NA |
```

Essentially:
* On 100 and 1000 iterations we are getting 1% to 2% improvement.
* On 10 iterations we are getting 5% loss.
* Why? The moveNext function is very consistent. The size of the array is unchanged. So for size 10, you get 9 fails, then one hit. The branch predictor is being smart and predicts this.
* For size 100 and 1000 the branch predictor isn't as good at going that wide, and the csel is slightly better.

Reply:
New version:
* SSA changes (including fixing the void)
* Cost calculations. I'm using the size here, as that's what we're concerned about. Oddly, calling gtSetEvalOrder() changes the previously calculated values, so I run it again to make sure everything is correct. Also, there is nothing taking into account that JTRUE(CMP(x,0)) would turn into cbnz x, so had to explicitly check for it here.